README.md
Odt2Xhtml
=========

Odt2Xhtml is a class PHP5 to transform files in format ODT to files (x)html.

Author
======

Stéphane HUC : <devs@stephane-huc.net>
GIT : https://github.com/hucste/Odt2Xhtml

Use-it
======

Please, see http://odt2xhtml.eu.org/index_en.html

Enjoy-it
========


CHANGELOG
CHANGELOG

=== 20110318 ===
Few modifications on code PHP

=== 20110315 ===
Lib PCLZip changed to version 2.8.2

=== 20110310 ===
The class is now rewriting in PHP5 POO...

=== 20090209 ===
New file : php5/mssg.php5 : class message to displaying in lib odt2xhtml

Modify odt2xhtml.php5 script :
- Add method to copy icon in directory html/img
- modify method _unzip_odf to function with ZIP support native on PHP 5
	(usage PCLZIP can disparate!)

PCLZIP is archived!

=== 20090202 ===
Add a capacity to use the library in PHP CLI. See test_cli.php5 script.

Correct file odt2xhtml.xsl.

=== 20090130 ===
Add a possibility to obtain css in file .css separate.

=== 20090119 ===
Modify odt2xhtml.php5 script :
- Add management of abbreviations (OOo specification text:reference-mark)
- modify display differents messages : method _message($value,$index,$var='')

Modify odt2xhtml.xsl :
- Add management element html abbr by using text:reference-mark : good ?

New version : odt2xhtml.odt

=== 20090118 ===
Add file .htaccess_for_php5.
Add file config.php5 :

Correct index.php5 script :
- modify absolute directory for files included

Correct odt2xhtml.php5 script :
- modify method _mk_all_dir() if making directories necessaries not run
- modify debuging messages.
- modify management of file and extension

Correct odt2xhtml.xsl : (support to office version 1.2)
- add element link to display icon
- add support to automatic styles in document style
- modify longdesc support in element img

Thanks to Damien Accorsi <damien AT accorsi POINT info>

=== 20080121 ===
Correct odt2xhtml.php5 script :
- error in method get_elements_html() : bad directory to read file html.

Thanks to Eric Bellot <contact AT eric-bellot POINT fr>

=== 20071122 ===
Modify XSL File :
- Resolve the bug in XSL that declarate 2 xmlns in element html.

Correct odt2xhtml.php5 script.

=== 20071121 ===
Remove script in PHP4 - official support stop 20071231 -
Modify script odt2xhtml.php5 : 
- reorganize it !
- new method to make all directories necessary : _mk_all_dir()

=== 20070731 ===
Modify script odt2xhtml.php(5) : 
- to manage text position indice or exposant
- to manage header and footer page
- to manage too files .sxw and .stw (StarOffice & OOo 1.0)
- new method get_elements_html() and display_elements_html() to include elements
   html converted by this script, in your file html.
   
Create new file XSLT sxw2xhtml.xsl to manage file .sxw and .stw.
Modify odt2xhtml.xsl to :
- correct the output html, the meta tags, output img segun is in h or p elements,
- implement div header, div footer, element sup and sub, blockquote citations,

=== 20070307 ===
Use another template xsl : odt2xhtml.xsl - (project ODFTools : http://opendocumentfellowship.org/projects/odftools)
Better styles supports by this file xsl.
I've modified and corrected the 'odt2xhtml.xsl' for build an ouput file html, W3C compliant, XHTML 1.0.

Modify script odt2xhtml.php(5) : 
- reorganization
- rename method private __exists_odt in _make_new_xml_with_odf_files
- modify method private __make_new_xml_with_odf_files to manage meta.xml, content.xml, styles.xml

=== 20070626 ===
Modify script odt2xhtml.php(5) : (necessary to run correctly with odt2xhtml_portable)
 - modify constructor to link correctly img ...
 - modify private method __delete_tmp() to public method delete_tmp ... 

Modify script test.php(5).

Correct script index.php(5) to run correctly on WINNT.

=== 20070625 ===
Modify the script index.php(5) to run on WINNT.

=== 20070621 ===
Implement versions PHP 4 and 5, separatly and correctly.
add file odt2xhtml.odt, to test the script.
writing script test.php(5?) 

=== 20061029 ===
- odt2xhtml.php
modify the script to run with php5.

*** 20060925 ***
delete script config.php

- odt2xhtml.php :
modify constructor to permit that script run with different paths

*** 20060915 ***
add script config.php !

- template.xsl : 
modify out element paragraph
modify out element anchor empty

- odt2xhtml.php : 
modify private method __exists_odt();
modify private method __move_img();

*** 20060902 ***
add define image directory
add public method delete_dir();
modify private method __move_img();

*** 20060825 ***
* First version


test.php
<?php
/***
 Just to test-it !

* Verify value ODT2XHTML_PHPCLI is egual O in file config *
***/

require_once('index.php');

if(ODT2XHTML_PHPCLI !== true) {
	
	define('ODT2XHTML_ROOT', dirname(__FILE__) );
	define('ODT2XHTML_FILE_ODF', 'odt2xhtml.sxw');	// name file with extension .odt, .ott or .sxw, .stw!
	define('ODT2XHTML_FRONTEND', dirname(__FILE__).'/');	// directory where file odt to converse
	
	$name_html = substr(ODT2XHTML_FILE_ODF, 0, -4).'.html';
	$file = ODT2XHTML_FRONTEND.'html/'.$name_html;
	$file_html = 'html/'.$name_html;

	if(file_exists($file) && is_file($file)) unlink($file);

	require_once(ODT2XHTML_ROOT.'/class/class.odt2xhtml.php');
	$class = new odt2xhtml(); 	// ODT2XHTML_ROOT,ODT2XHTML_FRONTEND,ODT2XHTML_FILE_ODT
	$class->convert2xhtml();

	if(file_exists($file) && is_file($file) && filesize($file) != 0) {
		echo '<p>Good test: <a href="'.$file_html.'">'.$name_html.'</a></p>';
	
		/*** if you desire to include code ODF2html in your file HTML ***/
		//$class->get_containers_html();
		## For use the method display_container_html()
		# you can use 'meta','css','body','title' 
		#  and :
		#   0 - to obtain elements with the containers
		#	1 - to obtain elements without the containers 
		#		(for all containers, but not 'title'), 
		#		or obtain class name (only for the container 'title')
		#	2 - to obtain elements without the container
		#		(only for the container 'title')
		//echo '<p>Here is a test to display container body :</p>';
		//$class->display_container_html('body', 1); 
		//echo '<p>Here is a test to display container css :</p>';
		//$class->display_container_html('css', 1); 
		//echo '<p>Here is a test to display container meta :</p>';
		//$class->display_container_html('meta', 1); 
		//echo '<p>Here is a test to display container body :</p>';
		//$class->display_container_html('body', 1); 
	}
	else echo '<p style="color:red; font-weight: bold; font-size: 2em;">Converse ODT to XHTML: fault!</p>';

	unset($file_html,$name_html,$class);
}
else die('<p style="color:red; font-weight: bold; font-size: 2em;">Modify the config file to set the variable ODT2XHTML_PHPCLI to "0"!</p>');
?>


test_cli.php
#!/usr/bin/php
<?php
/***
* Just to test-it !
*
* Think to modify value ODT2XHTML_PHPCLI at 1 in file index.php *
* Think to chmod +x test_cli.php
** To try it : ./test_cli.php /directory_in/file.odt /directory_out **
***/

require_once('index.php');

if(ODT2XHTML_PHPCLI == TRUE) {
	$mssg = "Specify argument as: 'php test_cli.php (/dir_in/)file.odt /dir_out'";
	switch($argc) {
		case 0:
		case 1: die("\033[31m".'*** STOP, here. No Argument specified! '.$mssg.' ***'."\033[37m\r\n");
		break;
		case 2: die("\033[31m".'*** Fault one argument; '.$mssg.' ***'."\033[37m\r\n");
		break;
	}
	if($argc>3) die("\033[31m".'*** Too much arguments: '.$mssg.'! ***'."\033[37m\r\n");

	if(!empty($argv[1])) {
		$x = array_reverse(explode('/', $argv[1]));
		
		# name file with extension .odt, .ott or .sxw, .stw!
		if(!empty($x[0])) $file = $x[0];
		
		if(file_exists($file) && is_file($file)) {
			
			# directory where file odt is 
			$root = str_replace($file,'',$argv[1]);
			
			if(empty($root)) $root = getcwd().'/';
			
			define('ODT2XHTML_ROOT', $root);
			# directory where file odt is out
			$frontend = $argv[2];
			
			# if ultimate caracter of $frontend isn't an '/'
			if(file_exists($frontend) && is_dir($frontend)) {
				if(substr($frontend, -1, 1) != '/') define('ODT2XHTML_FRONTEND', $frontend.'/');
				else define('ODT2XHTML_FRONTEND', $frontend);
			}
			
			define('ODT2XHTML_FILE_ODF', $file);

			$name_html = substr($file, 0, strrpos($file, ".")).'.html';
			$file_html = ODT2XHTML_FRONTEND.'html/'.$name_html; 

			if(file_exists($file_html) && is_file($file_html)) unlink($file_html);
	
			require_once(ODT2XHTML_ROOT.'class/class.odt2xhtml.php');
			$class = new odt2xhtml();
			$class->convert2xhtml();

			if(file_exists($file_html) && is_file($file_html) && filesize($file_html) != 0) {
				echo "\033[32m".'Good test!'."\033[37m".' file is on: '."\033[34m".$file_html."\033[37m\r\n";
			
				/*** if you desire to include code ODF2html in your file HTML ***/
				//$class->get_containers_html();
				## For use the method display_elements_html()
				# you can use 'meta','css','body','title' 
				#  and :
				#   0 - to obtain with the elements containers
				#	1 - to obtain without the elements containers
				//echo '<p>Here is a test to display elements css :</p>';
				//$class->display_container_html('css', 1); 
			}
			else echo "\033[31m".'*** Converse fault!***'."\033[37m\r\n";
			
			unset($file_html, $name_html, $frontend, $file, $class);
		}
	}
}
else die('<p style="color:red;">Modify the file index to set the variable ODT2XHTML_PHPCLI to "1"!</p>');
?>


config.php
<?php
/***
 * class PHP odt2xhtml : file config
 * Copyright (C) 2011  Stephane HUC
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Contact information:
 *   Stephane HUC
 *   <devs@stephane-huc.net>
 *
 ***/
define('ODT2XHTML_FILE_CSS', 1);	// to obtain css in file .css

define('ODT2XHTML_PUB', ' :: converted by Odt2Xhtml :: http://odt2xhtml.eu.org');	// message publicis
define('ODT2XHTML_TITLE', 'element_title');	// file_name or element_title

?>


class/class.messaging.php
<?php
/***
 * class PHP display_message : to display message in class odt2xhtml
 * Copyright (C) 2011  Stephane HUC
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Contact information:
 *   Stephane HUC
 *   <devs@stephane-huc.net>
 *
 ***/
/***
 * 
 * Use this class: 
 * 
 * $value = 'ok'; 	// value is 'ok', or 'ko'
 * $index = 'index_key_mssg'; 	// index is a key in files mssg 
 * $info = ''; 	// all information that you desire to displaying
 * 
 * $this->mssg = new messaging();
 * 
 * $this->mssg->display($value, $index, '');
 * 
 * */

define('CLASS_MSSG_ROOT', dirname(__FILE__));
define('CLASS_MSSG_LANG', 'en');
define('CLASS_MSSG_FILES_LANG', CLASS_MSSG_ROOT.'/lang/'.CLASS_MSSG_LANG.'/');

if(!defined('CLASS_MSSG_PHPCLI')) define('CLASS_MSSG_PHPCLI', 0);


class messaging {
	
	private $content;
	private $ko;
	private $ok;
	private $mssg;

	public function __construct() {
		
		$this->file_mssg = array (
			'ko.cfg',
			'ok.cfg',
		);
		
		$this->get_files_need();
		$this->set_variables();

	}

	public function __destruct() {
		unset($this->mssg);
	}
	
	/***
		$value 
		$index
		$var : is variable to obtain in method
	***/
	public function display($value, $index, $info='') {
		try {
			
			if(CLASS_MSSG_PHPCLI == true) {
				$color = array (
					'blue' => "\033[34m",
					'green' => "\033[32m",
					'red' => "\033[31m",
				);
			}
			
			switch($value) {
				case 'ko' :
					$this->mssg = '<p>';
					
					if(!empty($info)) {
						$this->mssg .= $this->ko[$index];
						
						if(CLASS_MSSG_PHPCLI == true) $this->mssg .= ' '.$color['blue'].$info."\033[37m ";
						else $this->mssg .= ' (<strong>'.$info.'</strong>)';
					}
					else $this->mssg .= $this->ko[$index];
					
					if(CLASS_MSSG_PHPCLI == true) $this->mssg .= ' '.$color['red'].'KO'."\033[37m ";
					else $this->mssg .= ' <strong style="color:red;">KO</strong>!</p>';
				break;
				
				case 'ok' :
					$this->mssg = '<p>';
					
					if(!empty($info)) {
						$this->mssg .= $this->ok[$index];
						
						if(CLASS_MSSG_PHPCLI == true) $this->mssg .= ' '.$color['blue'].$info."\033[37m ";
						else $this->mssg .= ' (<strong>'.$info.'</strong>)';
					}
					else $this->mssg .= $this->ok[$index];
					
					if(CLASS_MSSG_PHPCLI == true) $this->mssg .= ' '.$color['green'].'OK'."\033[37m ";
					else $this->mssg .= ' <strong style="color:green;">OK</strong>!</p>';
				break;
			}
		
			if(CLASS_MSSG_PHPCLI == TRUE) return strip_tags($this->mssg)."\n";
			else return $this->mssg."\n";
			
		}
		catch(Exception $e) { die('Error_method: '.__METHOD__.$e->getMessage()); }
	}
		
	private function get_files_need() {
		try {
			
			require_once(CLASS_MSSG_ROOT.'/class.file_handling.php');
			
			if(!empty($this->file_mssg) && is_array($this->file_mssg)) {
				foreach($this->file_mssg as $key => $value) {

					$var = substr($value, 0, 2); 

					$this->file_name = CLASS_MSSG_FILES_LANG.$value; 
					
					$this->file = new file_handling(); 
					if($this->file->verif_file($this->file_name)) {
						$this->content[$var] = $this->file->file_object($this->file_name, 'fgets'); 	
					}
					
				}
				unset($key, $value); 				
			}
			
		}
		catch(Exception $e) { die('Error_method: '.__METHOD__.$e->getMessage()); }
	}
	
	private function set_variables() {
		try {
			
			if(!empty($this->content)) {
				foreach($this->file_mssg as $key => $value) {
					$var = substr($value, 0, 2);
					
					foreach($this->content[$var] as $key1 => $value1) {
					
						if(!empty($value1)) {
							if(strpos($value1, '#') === false) {
								unset($this->content[$var][$key1]);
						
								$x = explode('=', $value1);
						
								if(is_array($x)) {
									$k = trim($x[0]);
									$v = trim($x[1]);
									$v = htmlentities($v, ENT_QUOTES);
									$this->content[$var][$k] = $v;
								}
								unset($x, $k, $v);
							}
							else unset($this->content[$var][$key1]);
						}
						else unset($this->content[$var][$key1]);
						
						if(!empty($this->content[$var])) {
							$$var = $this->content[$var];
						}
						
					}
					unset($key1, $value1); 
				}
				
				if(!empty($ok)) $this->ok = $ok;
				if(!empty($ko)) $this->ko = $ko;
				
			}
			
		}
		catch(Exception $e) { die('Error_method: '.__METHOD__.$e->getMessage()); }
	}
		
}
?>


class/class.file_handling.php
<?php
/***
 * class PHP display_message : to display message in class odt2xhtml
 * Copyright (C) 2011  Stephane HUC
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Contact information:
 *   Stephane HUC
 *   <devs@stephane-huc.net>
 *
 ***/
 
/***
 *
 * Use the class: 
 * 
 * $this->file = new file_handling();
 * 
 * if($this->file->verif_file($file)) $this->file->file_object($file);
 * 
***/ 
class file_handling {
	
	private $extensions=array();
	
	public function __construct() {
	}
	
	public function __destruct() {		
	}
	
	public function dir_object($path) {
		try {
			
			$this->it = new DirectoryIterator($path);
			
			return $this->it;
			
		}
		catch(Exception $e) { die('Error_method: '.__METHOD__.' '.$e->getMessage()); }
	}
	
	/*** obtain file info ***/
	public function file_infos($file, $info, $mode='', $resource='') {
		try {
			
			if(!empty($info) && is_string($info)) {
				switch($info) {
				
					case 'file_get_contents':
						return file_get_contents($file);
					break;
				
					case 'finfo':
						if(!empty($mode) && is_string($mode)) {
							switch($mode) {
						
								case 'mime':
									$this->finfo = finfo_open(FILEINFO_MIME);
								break;
						
								case 'mime_encoding':
									$this->finfo = finfo_open(FILEINFO_MIME_ENCODING);
								break;
								
								case 'mime_type':
									$this->finfo = finfo_open(FILEINFO_MIME_TYPE);
								break;
								
							}
							
							if(!empty($this->finfo)) {
								$this->mime = finfo_file($this->finfo, $file);
								finfo_close($this->finfo);
							}
							
							if(!empty($this->mime)) return $this->mime;
						}
					break;
				
					case 'pathinfo':
						$this->pathinfo = pathinfo($file);
					
						if(!empty($mode) && is_string($mode)) {
							switch($mode) {
				
								case 'base': 	return $this->pathinfo['basename']; 	break;
							
								case 'dir': 	return $this->pathinfo['dirname']; 	break;
						
								case 'ext': 	return $this->pathinfo['extension']; 	break;
						
								case 'name': 	return $this->pathinfo['filename']; 	break;
						
								default: 	return $this->pathinfo; 	break;
				
							}
						}
					break;
				
					case 'write':
						if( !$this->file->isFile() ||
							$this->file->isFile() && $this->file->isWritable() ) {
							
							$handle = $this->file->openFile($mode); 	
							$handle->fwrite($resource);
							//fclose($handle);

						}
					break;
				
				}
			
				unset($file);
			}
			
		}
		catch(Exception $e) { die('Error_method: '.__METHOD__.' '.$e->getMessage()); }
	}
	
	/*** obtain file content ***/
	public function file_object($file, $info='', $mode='', $resource='') {
		try {
			
			$this->file = new SplFileObject($file);
				
			switch($info) {
				
				case 'fgets':
					while(!$this->file->eof()) { $lines[] = trim($this->file->fgets()); }
					$this->lines = $lines; 
					return $this->lines;
				break;
				
				case 'write': 
					if( $this->file->isFile() && $this->file->isWritable() ) { 
						$handle = $this->file->openFile($mode); 	
						$handle->fwrite($resource);
					}
				break;
				
				default:
					return $this->file;
				break;
			
			}
			
			unset($file);
			
		}
		catch(Exception $e) { die('Error_method: '.__METHOD__.' '.$e->getMessage()); }
	}
	
	private function set_variables() {
		try {
			
			
			
		}
		catch(Exception $e) { die('Error_method: '.__METHOD__.' '.$e->getMessage()); }
	}
		
	/*** verif file ***/
	public function verif_dir($file) {
		try {
			
			$this->file = new SplFileInfo($file);
			if( !$this->file->isDir() 
				|| ( $this->file->isDir() && !$this->file->isReadable() )
				|| ( $this->file->getOwner() != fileowner(__FILE__) ) ) 
			{
				return FALSE;
			}
			else return TRUE;
			
		}
		catch(Exception $e) { die('Error_method: '.__METHOD__.$e->getMessage()); }
	}
	
	/*** verif file ***/
	public function verif_file($file) {
		try {
			
			$this->file = new SplFileInfo($file);
			if( !$this->file->isFile() 
				|| ( $this->file->isFile() && !$this->file->isReadable() )
				|| ( $this->file->getOwner() != fileowner(__FILE__) ) ) 
			{
				return FALSE;
			}
			else return TRUE;
			
		}
		catch(Exception $e) { die('Error_method: '.__METHOD__.$e->getMessage()); }
	}
	
 }
?>


index.php
<?php
/***
 * class PHP odt2xhtml : file index ... necessary to include !
 * Copyright (C) 2011  Stephane HUC
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Contact information:
 *   Stephane HUC
 *   <devs@stephane-huc.net>
 *
 ***/
 
define('ODT2XHTML_DEBUG', 0);
/*** to use with PHP CLI *** USE WITH PRECAUTION, @ YOURS RISKS ***/
define('ODT2XHTML_PHPCLI', 0);	
/*** get owner ***/
define('ODT2XHTML_OWNER', fileowner(__FILE__) );

/*** NOT TOUCH ***/
$hosted = array (
	'odt2xhtml.dev.stephane-huc.net',
	'odt2xhtml.eu.org',
);

$included = array (
	'config.php',
);

if(!empty($_SERVER['TERM']) && $_SERVER['TERM'] == 'xterm') {
	// if this script is execute on terminal (PHP CLI)
	error_reporting(E_ALL);
}
elseif(!empty($_SERVER['HTTP_HOST']) && in_array($_SERVER['HTTP_HOST'], $hosted)) {
	error_reporting(E_ALL);
}
else error_reporting(0);

if(class_exists('XSLTProcessor')) {
	
	foreach($included as $v) {
		$file = dirname(__FILE__).'/'.$v;
		
		$file = new SplFileInfo($file);
		if($file->isFile() && $file->isReadable() && ($file->getOwner() == ODT2XHTML_OWNER) ) {
			require_once($file);
			
			if(ODT2XHTML_DEBUG == true) { 
				$mssg = 'File included: ';
				
				if(ODT2XHTML_PHPCLI == true) {
					echo $mssg." \033[34m".$file."\033[37m \033[32m".'OK'."\033[37m\r\n";
				}
				else echo '<p style="color: green; font-weight: bold;">'.$mssg.'<strong>'.$file.'</strong> !</p>';
			}
		}
		else {
			$mssg = 'Failed to require file include needed: ';
			
			if(ODT2XHTML_PHPCLI == true) {
				die($mssg." \033[34m".$file."\033[37m \033[31m".'KO'."\033[37m\r\n");
			}
			else die('<p style="color:red; font-weight: bold; font-size: 2em">'.$mssg.$file.' !</p>');
		}
		
		unset($file); 
	}
	unset($v);	
}
else {
	$mssg = 'Your WebHosting not support correctly the XSL Transformation in PHP5! Needed PHP-XSL Library ...';
	
	if(ODT2XHTML_PHPCLI == true) {
		die("\033[31m".$mssg."\033[37m ");
	}
	else die('<p style="color:red;">'.$mssg.'</p>');
}
?>
